pkg/history/history.go

```diff
@@ -0,0 +1,409 @@
+// Enforces maximum entry limit
+func saveLogs(logs []CommandLog) error {
+	if len(logs) > maxEntries {
+		logs = logs[len(logs)-maxEntries:]
+	}
+
+	logPath, err := getLogPath()
+	if err != nil {
+		return fmt.Errorf("failed to get log path: %w", err)
+	}
+
+	data, err := json.MarshalIndent(logs, "", "  ")
+	if err != nil {
+		return fmt.Errorf("failed to marshal logs: %w", err)
+	}
+
+	if err := os.WriteFile(logPath, data, 0644); err != nil {
+		return fmt.Errorf("failed to write log file: %w", err)
+	}
+
+	return nil
+}
+
+// deriveCommandType determines the type of command being executed
+// Returns uppercase string representation of command hierarchy
+func deriveCommandType(cmd *cobra.Command) string {
+	path := cmd.CommandPath()
+	parts := strings.Fields(path)
+	if len(parts) <= 1 {
+		return "ROOT"
+	}
+	parts = parts[1:]
+	return strings.ToUpper(strings.Join(parts, "_"))
+}
+
+// appendLog adds a new command entry to the history
+func appendLog(entry CommandLog) error {
+	logs, err := loadLogs(0)
+	if err != nil {
+		logs = []CommandLog{}
+	}
+
+	logs = append(logs, entry)
+	return saveLogs(logs)
+}
+
+// wrapCommand wraps a cobra command to capture its execution details
+// Handles command interruption, output capture, and error logging
+func wrapCommand(cmd *cobra.Command) {
+	// Setup signal handling for interrupts
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
+
+	// Store original command functions
+	originalHelpFunc := cmd.HelpFunc()
+	originalPreRun := cmd.PreRun
+	originalPreRunE := cmd.PreRunE
+
+	// Wrap help function to capture help command execution
+	cmd.SetHelpFunc(func(cmd *cobra.Command, args []string) {
+		var stdout, stderr bytes.Buffer
+		oldOut := cmd.OutOrStdout()
+		oldErr := cmd.ErrOrStderr()
+
+		// Setup output capturing
+		cmd.SetOut(io.MultiWriter(oldOut, &stdout))
+		cmd.SetErr(io.MultiWriter(oldErr, &stderr))
+
+		// Run help in goroutine to handle interrupts
+		done := make(chan bool, 1)
+		go func() {
+			originalHelpFunc(cmd, args)
+			done <- true
+		}()
+
+		// Wait for completion or interrupt
+		select {
+		case <-done:
+		case <-sigChan:
+		}
+
+		// Log help command execution
+		logEntry := CommandLog{
+			Date:        time.Now(),
+			Input:       strings.Join(os.Args, " "),
+			Output:      stdout.String(),
+			Error:       stderr.String(),
+			ExitCode:    0,
+			CommandType: deriveCommandType(cmd),
+		}
+
+		if err := appendLog(logEntry); err != nil {
+			fmt.Fprintf(os.Stderr, "Warning: Failed to log help command: %v\n", err)
+		}
+	})
+
+	// Preserve PreRun functionality
+	cmd.PreRunE = func(cmd *cobra.Command, args []string) error {
+		if originalPreRunE != nil {
+			if err := originalPreRunE(cmd, args); err != nil {
+				return err
+			}
+		} else if originalPreRun != nil {
+			originalPreRun(cmd, args)
+		}
+		return nil
+	}
+
+	// Handle RunE commands (commands that can return errors)
+	if cmd.RunE != nil {
```

Comment:
Could we try to dry up the function bodies for `RunE` vs `Run` they seem to do ~80% of the same thing